src/models.py

```diff
@@ -45,38 +98,78 @@ def __init__(self, num_hidden, embedding_dim, vocab_size, device, num_layers=2,
         self.device = device
         
         self.emb = nn.Embedding(vocab_size, embedding_dim)
-        self.lstm = nn.LSTM(embedding_dim, num_hidden, num_layers=2, batch_first=True, dropout=0, bidirectional=bidirectional)
-        self.fc = nn.Linear(num_hidden, vocab_size)
+        self.lstm = nn.LSTM(embedding_dim, num_hidden, num_layers=2, batch_first=True, dropout=0, bidirectional=bidirectional) 
+        self.attn = nn.Linear(self.num_hidden * 2, 1) 
+        self.fc = nn.Linear(self.num_hidden * 2, vocab_size)
     
-    def forward(self, x, h0, seq_lengths):
-        # x has shape (batch, seq_len), h0 has shape (num_layers * num_directions, batch, num_hidden)
+    def forward(self, caption, h0, c0, seq_lengths, intermediate_layer):
+        # -------------- Inputs --------------------
+            # caption = caption; shape = (batch, seq_len)
+            # mlp = image feature vector averaged and passed through MLP
+            # seq_lengths = ''
+            # intermediate_layer = output of 'encoder.layer4[2].bn2' 
         
-        c0 = self.init_cell_state(h0.shape[1])
-        
-        emb = self.emb(x)  # (batch, seq_len) -> (batch, seq_len, embedding_dim)
+        # Initialising the hidden state and cell state of the LSTM with 
+        # the two different MLPs
+
+        c0, h0 = self.init_cell_state_and_hidden_state(caption.shape[0], c0, h0)
         
+        emb = self.emb(caption)  # (batch, seq_len) -> (batch, seq_len, embedding_dim)
+     
         # Pack_padded_sequence so that padded items in the sequence won't be shown to the LSTM
         # See https://towardsdatascience.com/taming-lstms-variable-sized-mini-batches-and-why-pytorch-is-good-for-your-health-61d35642972e
         # We use seq_lengths - 1 because we do not want to count the last <END> token in each caption.
         emb = torch.nn.utils.rnn.pack_padded_sequence(emb, seq_lengths - 1, batch_first=True)
+
+        # [2, 38, 512]
+
+        # input of shape (seq_len, batch, input_size):
         
         out, _ = self.lstm(emb, (h0, c0))
 
         # Undo the packing operation
         out, _ = torch.nn.utils.rnn.pad_packed_sequence(out, batch_first=True)  # (batch, seq_len, num_directions * num_hidden)
         
-        logits = self.fc(out)
+        # ---------------- Attention -----------------------
+
+        # torch.Size([2, 512, 16]) -> torch.Size([2, 16, 512])
+        intermediate_layer = torch.transpose(intermediate_layer, 1, -1)
+        
+        attn_in = torch.zeros(out.size(0), out.size(1), out.size(2) * 2).to(self.device)
```

Comment:
What it looks like to me is that for `attn_in`, you want to take the shape of `out`, and add another dimension to it that can fit all the 16 combinations of hidden state + one of the 16 annotation vectors, i.e. `attn_in.shape = [batch, seq_len, 16, num_hidden * 2]`. Then given a batch number `b`, timestep `t`, along with the `k'th` annotation vector `ann` and the hidden state `h` at batch `b` and time `t`, you have `attn_in[b, t, k, :] =  torch.cat(ann, h)`. In code:

```
# intermediate_layer has shape [batch, n_annotations, num_hidden],
# out has shape [batch, seq_len, num_hidden].

n_annotations = intermediate_layer.size(1)
attn_in = torch.zeros(batch, seq_len, n_annotations, num_hidden * 2)
for b in range(batch):
   for t in range(seq_len):
      h = out[b, t]
      for k, ann in enumerate(intermediate_layer[b]):  # iterate over all annotation vectors, 16 in this example
         attn_in[b, t, k, :] = torch.cat(ann, h)
```

You should start with writing this out in a for loop as above, because it's easier to verify what you're doing this way. You should only consider making it more efficient by replacing (part of) the for loops with commands like `torch.cat` once you are sure that you understand the loopy version properly.